proposals/context.md

```diff
@@ -0,0 +1,265 @@
+# context-protocol
+
+An open protocol for data passing between components.
+
+Author: Benjamin Delarre
+
+Document status: Draft
+
+Last update: 2021-4-27
+
+# Background
+
+There are a number of scenarios where a web component might require data that is provided from outside itself. While components can specify properties and attributes to receive that data imperatively or declaratively, it is often the case that the data may be owned somewhere further up the DOM tree.
+
+One approach to passing this data down to components is commonly referred to as 'prop drilling', whereby components pass properties all the way down through the hierarchy, passing from component to component until it is consumed at its destination. This is generally considered undesirable as it often requires intermediate components in the tree to have knowledge of the data necessary in its descendents.
+
+Frameworks and libraries often provide mechanisms for this, these can range from the simple Context implementation available in React, to more complex Dependency Injection frameworks. Web components need a similar protocol in order to solve this problem.
+
+# Goals
+
+- Allow elements in the DOM to retrieve data based on their contextual position in the DOM
+- Alleviate the problem of 'prop drilling'
+- Simple API that is easily implemented in any framework / library
+- Synchronous protocol, while supporting asynchronous patterns
+- Support single or multiple delivery of context values
+
+# Non-Goals
+
+**Context API !== Dependency Injection Framework**
+
+The Context API does not intend to cover all cases and forms of Dependency Injection. It does not specify constructor, factory or property injection patterns. Its only goal is to formalize the pattern of sharing data across the hierarchy in the DOM, specifically avoiding 'prop drilling' type scenarios. Dependency Injection patterns could be implemented using this protocol, but this is not the goal and should remain explicitly outside the scope of Context API for simplicity.
+
+**Context API is not a state management alternative**
+
+State management libraries often need to perform similar behaviors to the problems that Context API helps to solve. An element deep in the DOM tree made need access to some state, and may need to respond to that state being changed. While state management could be built using the Context API, it is not a primary goal of Context API to solve this problem.
+
+# Details
```

Comment:
I think it's be nice to have a brief overview before the details, which describes at a very high level the approach of firing an event to represent a request for a context object, that the event carries a callback to provide values, and that any listener above the dispatcher in the tree can handle the event, provide an object, and stop propagation.

Reply:
Gave a bullet point list outline in bccaeac